1.

The decimal number, 585 = 10010010012 (binary), is palindromic in both bases. Find all numbers under 1000 that are palindromic in base 10 and base 2. (Please note that the palindromic number, in either base, may not include leading zeros.) From https://projecteuler.net/problem=36

In [3]:
for n in range (1,1001):
    string_base_2=''
    string_base_10=''
    i=n
    j=n
    
    while i != 0:        
        string_base_10 += str(i%10)
        i = i//10
    
    if string_base_10[::-1] == string_base_10:
        while j != 0:
            string_base_2 += str(j%2)
            j = j//2
        if string_base_2[::-1] == string_base_2:
            print(n,'is a palindromic number in base 10 and base 2.')

1 is a palindromic number in base 10 and base 2.
3 is a palindromic number in base 10 and base 2.
5 is a palindromic number in base 10 and base 2.
7 is a palindromic number in base 10 and base 2.
9 is a palindromic number in base 10 and base 2.
33 is a palindromic number in base 10 and base 2.
99 is a palindromic number in base 10 and base 2.
313 is a palindromic number in base 10 and base 2.
585 is a palindromic number in base 10 and base 2.
717 is a palindromic number in base 10 and base 2.


2.

Here is a 1000 digit number. 
73167176531330624919225119674426574742355349194934 
96983520312774506326239578318016984801869478851843 
85861560789112949495459501737958331952853208805511 
12540698747158523863050715693290963295227443043557 
66896648950445244523161731856403098711121722383113 
62229893423380308135336276614282806444486645238749 
30358907296290491560440772390713810515859307960866 
70172427121883998797908792274921901699720888093776 
65727333001053367881220235421809751254540594752243 
52584907711670556013604839586446706324415722155397 
53697817977846174064955149290862569321978468622482 
83972241375657056057490261407972968652414535100474 
82166370484403199890008895243450658541227588666881 
16427171479924442928230863465674813919123162824586 
17866458359124566529476545682848912883142607690042 
24219022671055626321111109370544217506941658960408 
07198403850962455444362981230987879927244284909188 
84580156166097919133875499200524063689912560717606 
05886116467109405077541002256983155200055935729725 
71636269561882670428252483600823257530420752963450 
 
Find the four adjacent digits in the 1000-digit number that have the greatest product. What are the four adjacent digits and what is the value of this product? 
From <https://projecteuler.net/problem=8> 

In [10]:
string = string.replace("\n","")
string = string.replace(" ","")

string=str("""73167176531330624919225119674426574742355349194934\
96983520312774506326239578318016984801869478851843\
85861560789112949495459501737958331952853208805511\
12540698747158523863050715693290963295227443043557\
66896648950445244523161731856403098711121722383113\
62229893423380308135336276614282806444486645238749\
30358907296290491560440772390713810515859307960866\
70172427121883998797908792274921901699720888093776\
65727333001053367881220235421809751254540594752243\
52584907711670556013604839586446706324415722155397\
53697817977846174064955149290862569321978468622482\
83972241375657056057490261407972968652414535100474\
82166370484403199890008895243450658541227588666881\
16427171479924442928230863465674813919123162824586\
17866458359124566529476545682848912883142607690042\
24219022671055626321111109370544217506941658960408\
07198403850962455444362981230987879927244284909188\
84580156166097919133875499200524063689912560717606\
05886116467109405077541002256983155200055935729725\
71636269561882670428252483600823257530420752963450""").strip() #.replace(" ","")
#print(string)
maximum=-1
max_str=''
for i in range (997): #[0:996]
    num=string[i:i+4]
    product=1
    for i in num:
        product *= int(i)

    if product > maximum:
        maximum = product
        max_str = num
        
print('The value of this product is:', maximum)
print('The four digits are:'+max_str[0]+','+max_str[1]+','+max_str[2]+','+max_str[3])

The value of this product is: 5832
The four digits are:9,9,8,9


3.

In [5]:
#What if there's a question mark in the middle of the whole sentence?
def capitalize_letter(s):
    sen=s.split('[. ')
    print(sen)
    string=''
    for i in sen:
        j=i[0].upper() + i[1:] + '. '
        string=string+j
    print(string[:-2])
    
    return()

def main():
    string=input("Enter a string: ")
    capitalize_letter(string)
    return

main()

Enter a string: i'm a dog. i'm a happy dog. i'm a happy happy dog.
["i'm a dog. i'm a happy dog. i'm a happy happy dog."]
I'm a dog. i'm a happy dog. i'm a happy happy dog.


Others:

Are the codes right?

A palindromic prime is a number that is simultaneously palindromic and prime. Find palindromic prime numbers under 1000. From http://mathworld.wolfram.com/PalindromicPrime.html

In [6]:
#What is the base of the palindromic number?
#decimal???
#The first few decimal palindromic primes are:
#2, 3, 5, 7, 11, 101, 131, 151, 181, 191, 313, 353, 373, 383, 727, 757, 787, 797, 919, 929, …
for i in range (1,1000):
    if str(i) == str(i)[::-1]:
        #then i is a  palindromic number.
        #find whether it's also a prime number or not
        is_prime=True
        if i < 2:
            is_prime=False
        else:
            for j in range (2,i):
                if i%j == 0:
                    is_prime=False
            if is_prime:
                print(i,'is a palindromic prime number')


2 is a palindromic prime number
3 is a palindromic prime number
5 is a palindromic prime number
7 is a palindromic prime number
11 is a palindromic prime number
101 is a palindromic prime number
131 is a palindromic prime number
151 is a palindromic prime number
181 is a palindromic prime number
191 is a palindromic prime number
313 is a palindromic prime number
353 is a palindromic prime number
373 is a palindromic prime number
383 is a palindromic prime number
727 is a palindromic prime number
757 is a palindromic prime number
787 is a palindromic prime number
797 is a palindromic prime number
919 is a palindromic prime number
929 is a palindromic prime number


If we take 47, reverse and add, 47 + 74 = 121, which is palindromic. Check for such numbers below 1000. 
From <https://projecteuler.net/problem=55> 

In [6]:
for i in range (1,1000):
    j=int(str(i)[::-1])
    h=i+j
    if str(h) == str(h)[::-1]:
        print(str(i)+'+'+str(j)+'='+str(h))

1+1=2
2+2=4
3+3=6
4+4=8
10+1=11
11+11=22
12+21=33
13+31=44
14+41=55
15+51=66
16+61=77
17+71=88
18+81=99
20+2=22
21+12=33
22+22=44
23+32=55
24+42=66
25+52=77
26+62=88
27+72=99
29+92=121
30+3=33
31+13=44
32+23=55
33+33=66
34+43=77
35+53=88
36+63=99
38+83=121
40+4=44
41+14=55
42+24=66
43+34=77
44+44=88
45+54=99
47+74=121
50+5=55
51+15=66
52+25=77
53+35=88
54+45=99
56+65=121
60+6=66
61+16=77
62+26=88
63+36=99
65+56=121
70+7=77
71+17=88
72+27=99
74+47=121
80+8=88
81+18=99
83+38=121
90+9=99
92+29=121
100+1=101
101+101=202
102+201=303
103+301=404
104+401=505
105+501=606
106+601=707
107+701=808
108+801=909
110+11=121
111+111=222
112+211=323
113+311=424
114+411=525
115+511=626
116+611=727
117+711=828
118+811=929
120+21=141
121+121=242
122+221=343
123+321=444
124+421=545
125+521=646
126+621=747
127+721=848
128+821=949
130+31=161
131+131=262
132+231=363
133+331=464
134+431=565
135+531=666
136+631=767
137+731=868
138+831=969
140+41=181
141+141=282
142+241=383
143+341=484
144+441=585
145+541=686
14

145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145. That is, the sum of the factorial of each of its digits equals the number itself. Write code to display all numbers below 1000 for which this is true. (30 points)

In [ ]:
#There only have one curious number under 1000?
#Is there any better way to solve this problem?
def sum_of_factorial(n):
    total=1
    for x in range (1,n+1):
        total*=x
    return total

def find_curious_number():
    i=1
    j=1
    h=1
    while(sum_of_factorial(i)+sum_of_factorial(j)+sum_of_factorial(h) < 1000):
        for i in range (1,1000):
            for j in range(1,1000):
                for h in range(1,1000):
                    summ=sum_of_factorial(i)+sum_of_factorial(j)+sum_of_factorial(h)
                    if str(summ) == str(i)+str(j)+str(h):
                        print(str(summ)+'='+str(i)+'!'+str(j)+'!'+str(h)+'!')
    return #str(summ)+'='+str(i)+'!'+str(j)+'!'+str(h)+'!'

find_curious_number()

145=1!4!5!


In [7]:
#Better solution
def sum_of_factorial(n):
    total=1
    if n == 0:
        return 1
    if n > 0:
        for x in range (1,n+1):
            total*=x
        return total

def main():
    for i in range(1000):
        mystring=str(i)
        mylength=len(mystring)
        total=0
        for j in range (mylength):
            mynumber=int(mystring[j])
            total += sum_of_factorial(mynumber)
        if mystring == str(total):
            print(i)                                      

main()

1
2
145
